lib/api.js

```diff
@@ -289,14 +289,19 @@ class API {
       // already initialized.
       await provider.init();
     }
-
-    let cur = provider;
-
+    
     // Go down to core HardhatNetworkProvider
+    let cur = provider;
     while (cur._wrapped) {
       cur = Object.assign({}, cur._wrapped)
     }
-    cur._node._vm.evm.events.on('step', self.collector.step.bind(self.collector))
+    
+    // Hardhat dual-mode node compatibility
+    if ('_context' in cur._node) {
+      cur._node._context._vm._vm.evm.events.on('step', self.collector.step.bind(self.collector))
+    } else {
+      cur._node._vm.evm.events.on('step', self.collector.step.bind(self.collector))
```

Comment:
Hmm, this change shouldn't have any impact (except for the extra `if`) right now, because the `else` branch would only be used with a version of Hardhat that's not out there yet.